sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+            uart_ttys,
+            inner: Rc::new(Inner {
+                console_tty: console_tty.ok_or(Error::CommunicationError("Missing console interface"))?,
+                usb_device: RefCell::new(device),
+                gpio: Default::default(),
+                spis: Default::default(),
+                uarts: Default::default(),
+            })
+        };
+        Ok(result)
+    }
+
+    /// Locates the /dev/ttyUSBn node corresponding to a given interface in the sys directory
+    /// tree, e.g. /sys/bus/usb/devices/1-4/1-4:1.0 .
+    fn find_tty(path: &Path) -> Result<PathBuf> {
+        for entry in fs::read_dir(path)? {
+            let dir_entry = entry?;
+            match dir_entry.file_name().into_string() {
+                Ok(filename) => {
+                    if filename.starts_with("tty") {
+                        return Ok(PathBuf::from("/dev").join(dir_entry.file_name()));
+                    }
+                }
+                _ => ()
+            }
+        }
+        Err(Error::CommunicationError("Did not find ttyUSBn device").into())
+    }
+}
+
+impl Inner {
+
+    /// Send a command to HyperDebug firmware, with a callback to receive any output.
+    pub fn execute_command(
+        &self,
+        cmd: &str,
+        mut callback: impl FnMut(&str)) -> Result<()>
+    {
+        use std::io::Read;
+        use std::io::Write;
+        let mut port = serialport::new(self.console_tty.to_str().ok_or(Error::UnicodePathError)?, 115_200)
+            .timeout(std::time::Duration::from_millis(100))
+            .open().expect("Failed to open port");
+
+        // Ideally, we would invoke Linux flock() on the serial
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf = [0u8; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    log::info!("Discarded {} characters: {:?}\n", rc,
+                               &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            break;
+                        }
+                        _ => {
+                            return Err(error.into())
+                        }
+                    }
+                }
+            }
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
+        let mut seen_echo = false;
+        let mut len: usize = 0;
+        loop {
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
```

Comment:
Could you describe what this algorithm does? Perhaps I can figure out a more succinct way to phrase it.

Reply:
The USB device has basically a console on ttyUSBn, that you could connect to, type commands, and have it do various things.  The USB device supports a very simple command editor, which includes among other things: echo of input characters, and a prompt character. This driver is connecting to that USB interface and could be sending a string like "gpioget A6\n", to which it is expecting a reply that could be "gpioget A6\n0 A6\n> ".
What code here is trying to do is first to break the input up by the newline characters, discard anything that precedes the echo of the command that it just sent, and then continue reading until it gets the two characters "> " at the beginning of a line, (and that the serial input read() times out right after that second character.)

Reply:
Sounds like this will do the trick:
```rust
// You may need to pull a newline off the end of cmd.
let mut split = buf.split(|b| b == b'\n').skip_while(|line| != cmd);
let _echo = split.next();
for line in split {
  if line.starts_with(b"> ") { break }
  callback(line);
}
```
That should be a starting point at a minimum.

Reply:
One issue is that for performance, I want to stop after a short timeout, if the last characters received are ```"> "```, but I want to employ a much longer timeout, if the last characters are anything else.  Some commands may take a few seconds to start producing output, or maybe pause in between lines.  So I do not think that I can separate the algorithm into one part, that gathers all the data into ```buf```, and a second part which breaks ```buf``` into lines.

Reply:
I have reworked the control structure, to avoid the ugly labelled outer loop, I hope this has helped make the logic easier to follow.